src/odemis/acq/align/z_localization.py

```diff
@@ -0,0 +1,193 @@
+    Function to determine the z position of feature in an image that was taken with a lens with astigmatism and
+    corresponding calibration data. Via a Gaussian fit the width and height of the feature are determined. Using a
+    fit on the equation of Huang the z position is then approximated. The function includes various warning flags
+    that may be raised when the calculation seems to give inaccurate results.
+
+    :param image (numpy.array): 2d array containing only the feature to be analyzed.
+    :param calibration_data (dict): contains the data from the calibration performed using the jupyter notebook with the
+                                    following keys:
+                                    x (dict) --> a, b ,c, d, w0 (floats), fit on the equation of Huang in x direction
+                                    y (dict) --> a, b ,c, d, w0 (floats), fit on the equation of Huang in y direction
+                                    feature_angle (float) angle of the ellipsoidal shapes w.r.t. the positive Y axis (anti-clockwise is positive)
+                                    upsample_factor (int) number with which the data is up sampled during the calibration
+                                    z_least_confusion (float) location in z in meters where the least confusion is present in the image (in focus)
+                                    z_calibration_range (tuple) --> (min, max) floats of the min and max z value in meters w.r.t. the z_least_confusion
+
+    :param z_stack_step_size (float): the step size in meters between two images in the z stack
+    :param fit_tol (float): factor for the precision of the Gaussian fit, range from 0 --> 1
+    :return:
+        z_position (float): determined z position of the feature in meter
+        warn_lvl (int): None = No warnings
+                        1 = from fmin_cg(scipy) max number of iterations exceeded
+                        2 = from fmin_cg(scipy) gradient and/or function calls were not changing
+                        3 = from fmin_cg(scipy) NaN result encountered
+                        4 = Outputted Z position is outside the defined maximum range, output is inaccurate
+                        5 = The Gaussian fit is not precise enough, probably because the image contains too much noise
+                        6 = The Gaussian fit found a feature to big for the current feature, the size > 85%
+    :raises:
+            ModuleNotFoundError if the module psf_extractor is not found.
+            KeyError if the calibration data is incomplete and does not include all the keys
+
+    """
+    warning = None  # Set the warning level to None, no warnings
+
+    if not psf_extractor:
+        raise ModuleNotFoundError("The psf_extractor module is not found. Cannot determine the Z position")
+
+    # Fit a Gaussian to the feature in the image
+    popt, pcov = psf_extractor.extractor.fit_gaussian_2D(image, theta=numpy.rad2deg(calibration_data["feature_angle"]), epsilon=1e-6)
```

Comment:
For single values, please use the standard Python function: `math.degrees(calibration_data["feature_angle"])`